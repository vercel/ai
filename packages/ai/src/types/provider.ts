import { EmbeddingModel } from './embedding-model';
import { LanguageModel } from './language-model';
import { ImageModel } from './image-model';

/**
 * Provider for language, text embedding, and image models.
 */
export type Provider = {
  /**
  Returns the language model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {LanguageModel} The language model associated with the id

  @throws {NoSuchModelError} If no such model exists.
     */
  languageModel(modelId: string): LanguageModel;

  /**
  Returns the text embedding model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {LanguageModel} The language model associated with the id

  @throws {NoSuchModelError} If no such model exists.
     */
  textEmbeddingModel(modelId: string): EmbeddingModel<string>;

  /**
  Returns the image model with the given id.
  The model id is then passed to the provider function to get the model.

  @param {string} id - The id of the model to return.

  @returns {ImageModel} The image model associated with the id
  */
  imageModel(modelId: string): ImageModel;
};
