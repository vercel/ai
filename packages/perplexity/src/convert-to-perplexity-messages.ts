import {
  LanguageModelV2Prompt,
  UnsupportedFunctionalityError,
} from '@ai-sdk/provider';
import { PerplexityPrompt } from './perplexity-language-model-prompt';

export function convertToPerplexityMessages(
  prompt: LanguageModelV2Prompt,
): PerplexityPrompt {
  const messages: PerplexityPrompt = [];

  for (const { role, content } of prompt) {
    switch (role) {
      case 'system': {
        messages.push({ role: 'system', content });
        break;
      }

      case 'user':
      case 'assistant': {
        messages.push({
          role,
          content: content
            .map(part => {
              switch (part.type) {
                case 'text': {
                  return part.text;
                }
              }
            })
            .filter(Boolean)
            .join(''),
        });
        break;
      }
      case 'tool': {
        throw new UnsupportedFunctionalityError({
          functionality: 'Tool messages',
        });
      }
      default: {
        const _exhaustiveCheck: never = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }

  return messages;
}
