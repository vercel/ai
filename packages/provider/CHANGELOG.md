# @ai-sdk/provider

## 1.0.11

### Patch Changes

- 2e1101a: feat (provider/openai): pdf input support

## 1.0.10

### Patch Changes

- e1d3d42: feat (ai): expose raw response body in generateText and generateObject

## 1.0.9

### Patch Changes

- ddf9740: feat (ai): add anthropic reasoning

## 1.0.8

### Patch Changes

- 2761f06: fix (ai/provider): publish with LanguageModelV1Source

## 1.0.7

### Patch Changes

- d89c3b9: feat (provider): add image model support to provider specification

## 1.0.6

### Patch Changes

- 3a58a2e: feat (ai/core): throw NoImageGeneratedError from generateImage when no predictions are returned.

## 1.0.5

### Patch Changes

- 0a699f1: feat: add reasoning token support

## 1.0.4

### Patch Changes

- 19a2ce7: feat (provider): add message option to UnsupportedFunctionalityError
- 6337688: feat: change image generation errors to warnings

## 1.0.3

### Patch Changes

- 5ed5e45: chore (config): Use ts-library.json tsconfig for no-UI libs.

## 1.0.2

### Patch Changes

- 09a9cab: feat (ai/core): add experimental generateImage function

## 1.0.1

### Patch Changes

- b446ae5: feat (provider): Define type for ObjectGenerationMode.

## 1.0.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- c0ddc24: chore (ai): remove toJSON method from AI SDK errors

## 1.0.0-canary.0

### Major Changes

- b469a7e: chore: remove isXXXError methods
- c0ddc24: chore (ai): remove toJSON method from AI SDK errors

## 0.0.26

### Patch Changes

- aa98cdb: chore: more flexible dependency versioning
- 1486128: feat: add supportsUrl to language model specification
- 7b937c5: feat (provider-utils): improve id generator robustness
- 3b1b69a: feat: provider-defined tools
- 811a317: feat (ai/core): multi-part tool results (incl. images)

## 0.0.25

### Patch Changes

- b9b0d7b: feat (ai): access raw request body

## 0.0.24

### Patch Changes

- d595d0d: feat (ai/core): file content parts

## 0.0.23

### Patch Changes

- 03313cd: feat (ai): expose response id, response model, response timestamp in telemetry and api
- 3be7c1c: fix (provider/anthropic): support prompt caching on assistant messages

## 0.0.22

### Patch Changes

- 26515cb: feat (ai/provider): introduce ProviderV1 specification

## 0.0.21

### Patch Changes

- f2c025e: feat (ai/core): prompt validation

## 0.0.20

### Patch Changes

- 6ac355e: feat (provider/anthropic): add cache control support

## 0.0.19

### Patch Changes

- dd4a0f5: fix (ai/provider): remove invalid check in isJSONParseError

## 0.0.18

### Patch Changes

- 4bd27a9: chore (ai/provider): refactor type validation

## 0.0.17

### Patch Changes

- 029af4c: feat (ai/core): support schema name & description in generateObject & streamObject

## 0.0.16

### Patch Changes

- d58517b: feat (ai/openai): structured outputs

## 0.0.15

### Patch Changes

- 96aed25: fix (ai/provider): release new version

## 0.0.14

### Patch Changes

- a8d1c9e9: feat (ai/core): parallel image download

## 0.0.13

### Patch Changes

- 2b9da0f0: feat (core): support stopSequences setting.
- a5b58845: feat (core): support topK setting
- 4aa8deb3: feat (provider): support responseFormat setting in provider api
- 13b27ec6: chore (ai/core): remove grammar mode

## 0.0.12

### Patch Changes

- b7290943: feat (ai/core): add token usage to embed and embedMany

## 0.0.11

### Patch Changes

- 5edc6110: feat (provider): add headers support to language and embedding model spec

## 0.0.10

### Patch Changes

- 102ca22f: fix (@ai-sdk/provider): fix TypeValidationError.isTypeValidationError

## 0.0.9

### Patch Changes

- 09295e2e: feat (@ai-sdk/provider): add DownloadError

## 0.0.8

### Patch Changes

- f39c0dd2: feat (provider): add toolChoice to language model specification

## 0.0.7

### Patch Changes

- 8e780288: feat (ai/provider): add "unknown" finish reason (for models that don't provide a finish reason)

## 0.0.6

### Patch Changes

- 6a50ac4: feat (provider): add additional error types

## 0.0.5

### Patch Changes

- 0f6bc4e: feat (ai/core): add embed function

## 0.0.4

### Patch Changes

- 325ca55: feat (ai/core): improve image content part error message

## 0.0.3

### Patch Changes

- 41d5736: ai/core: re-expose language model types.

## 0.0.2

### Patch Changes

- d6431ae: ai/core: add logprobs support (thanks @SamStenner for the contribution)
- 25f3350: ai/core: add support for getting raw response headers.

## 0.0.1

### Patch Changes

- eb150a6: ai/core: remove scaling of setting values (breaking change). If you were using the temperature, frequency penalty, or presence penalty settings, you need to update the providers and adjust the setting values.
