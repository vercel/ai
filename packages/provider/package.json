{
  "name": "@ai-sdk/provider",
  "version": "1.1.0",
  "license": "Apache-2.0",
  "sideEffects": false,
  "main": "./dist/index.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "files": [
    "dist/**/*",
    "CHANGELOG.md"
  ],
  "scripts": {
    "build": "tsup",
    "build:watch": "tsup --watch",
    "clean": "rm -rf dist",
    "lint": "eslint \"./**/*.ts*\"",
    "type-check": "tsc --noEmit",
    "prettier-check": "prettier --check \"./**/*.ts*\""
  },
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "dependencies": {
    "json-schema": "^0.4.0"
  },
  "devDependencies": {
    "@types/json-schema": "7.0.15",
    "@types/node": "20.17.24",
    "@vercel/ai-tsconfig": "workspace:*",
    "tsup": "^8",
    "typescript": "5.6.3"
  },
  "engines": {
    "node": ">=18"
  },
  "publishConfig": {
    "access": "public"
  },
  "homepage": "https://sdk.vercel.ai/docs",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vercel/ai.git"
  },
  "bugs": {
    "url": "https://github.com/vercel/ai/issues"
  },
  "keywords": [
    "ai"
  ]
}
