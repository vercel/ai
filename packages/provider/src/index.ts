export * from './embedding-model/index';
export * from './errors/index';
export * from './image-model/index';
export * from './json-value/index';
export * from './language-model/index';
export * from './provider/index';
export * from './transcription-model/index';

export type { JSONSchema7, JSONSchema7Definition } from 'json-schema';
