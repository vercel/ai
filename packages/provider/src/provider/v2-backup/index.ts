export type { ProviderV2 } from './provider-v2';
