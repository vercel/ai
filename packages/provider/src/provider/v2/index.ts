export type { ProviderV1 } from './provider-v1';
