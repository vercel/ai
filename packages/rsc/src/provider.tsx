// This file provides the AI context to all AI Actions via AsyncLocalStorage.

import * as React from 'react';
import { InternalAIProvider } from './rsc-shared.mjs';
import {
  withAIState,
  getAIStateDeltaPromise,
  sealMutableAIState,
} from './ai-state';
import type {
  ServerWrappedActions,
  AIAction,
  AIActions,
  AIProvider,
  InternalAIStateStorageOptions,
  OnSetAIState,
  OnGetUIState,
} from './types';

async function innerAction<T>(
  {
    action,
    options,
  }: { action: AIAction; options: InternalAIStateStorageOptions },
  state: T,
  ...args: unknown[]
) {
  'use server';
  return await withAIState(
    {
      state,
      options,
    },
    async () => {
      const result = await action(...args);
      sealMutableAIState();
      return [getAIStateDeltaPromise() as Promise<T>, result];
    },
  );
}

function wrapAction<T = unknown>(
  action: AIAction,
  options: InternalAIStateStorageOptions,
) {
  return innerAction.bind(null, { action, options }) as AIAction<T>;
}

export function createAI<
  AIState = any,
  UIState = any,
  Actions extends AIActions = {},
>({
  actions,
  initialAIState,
  initialUIState,

  onSetAIState,
  onGetUIState,
}: {
  actions: Actions;
  initialAIState?: AIState;
  initialUIState?: UIState;

  /**
   * This function is called whenever the AI state is updated by an Action.
   * You can use this to persist the AI state to a database, or to send it to a
   * logging service.
   */
  onSetAIState?: OnSetAIState<AIState>;

  /**
   * This function is used to retrieve the UI state based on the AI state.
   * For example, to render the initial UI state based on a given AI state, or
   * to sync the UI state when the application is already loaded.
   *
   * If returning `undefined`, the client side UI state will not be updated.
   *
   * This function must be annotated with the `"use server"` directive.
   *
   * @example
   * ```tsx
   * onGetUIState: async () => {
   *   'use server';
   *
   *   const currentAIState = getAIState();
   *   const externalAIState = await loadAIStateFromDatabase();
   *
   *   if (currentAIState === externalAIState) return undefined;
   *
   *   // Update current AI state and return the new UI state
   *   const state = getMutableAIState()
   *   state.done(externalAIState)
   *
   *   return <div>...</div>;
   * }
   * ```
   */
  onGetUIState?: OnGetUIState<UIState>;
}) {
  // Wrap all actions with our HoC.
  const wrappedActions: ServerWrappedActions = {};
  for (const name in actions) {
    wrappedActions[name] = wrapAction(actions[name], {
      onSetAIState,
    });
  }

  const wrappedSyncUIState = onGetUIState
    ? wrapAction(onGetUIState, {})
    : undefined;

  const AI: AIProvider<AIState, UIState, Actions> = async props => {
    if ('useState' in React) {
      // This file must be running on the React Server layer.
      // Ideally we should be using `import "server-only"` here but we can have a
      // more customized error message with this implementation.
      throw new Error(
        'This component can only be used inside Server Components.',
      );
    }

    let uiState = props.initialUIState ?? initialUIState;
    let aiState = props.initialAIState ?? initialAIState;
    let aiStateDelta = undefined;

    if (wrappedSyncUIState) {
      const [newAIStateDelta, newUIState] = await wrappedSyncUIState(aiState);
      if (newUIState !== undefined) {
        aiStateDelta = newAIStateDelta;
        uiState = newUIState;
      }
    }

    return (
      <InternalAIProvider
        wrappedActions={wrappedActions}
        wrappedSyncUIState={wrappedSyncUIState}
        initialUIState={uiState}
        initialAIState={aiState}
        initialAIStatePatch={aiStateDelta}
      >
        {props.children}
      </InternalAIProvider>
    );
  };

  return AI;
}
