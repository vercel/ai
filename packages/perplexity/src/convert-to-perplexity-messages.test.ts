import { convertToPerplexityMessages } from './convert-to-perplexity-messages';
import { UnsupportedFunctionalityError } from '@ai-sdk/provider';

describe('convertToPerplexityMessages', () => {
  describe('system messages', () => {
    it('should convert a system message with text content', () => {
      expect(
        convertToPerplexityMessages([
          {
            role: 'system',
            content: 'System initialization',
          },
        ]),
      ).toMatchSnapshot();
    });
  });

  describe('user messages', () => {
    it('should convert a user message with text parts', () => {
      expect(
        convertToPerplexityMessages([
          {
            role: 'user',
            content: [
              { type: 'text', text: 'Hello ' },
              { type: 'text', text: 'World' },
            ],
          },
        ]),
      ).toMatchSnapshot();
    });
  });

  describe('assistant messages', () => {
    it('should convert an assistant message with text content', () => {
      expect(
        convertToPerplexityMessages([
          {
            role: 'assistant',
            content: [{ type: 'text', text: 'Assistant reply' }],
          },
        ]),
      ).toMatchSnapshot();
    });
  });

  describe('tool messages', () => {
    it('should throw an error for tool messages', () => {
      expect(() => {
        convertToPerplexityMessages([
          {
            role: 'tool',
            content: [
              {
                type: 'tool-result',
                toolCallId: 'dummy-tool-call-id',
                toolName: 'dummy-tool-name',
                result: 'This should fail',
              },
            ],
          },
        ]);
      }).toThrow(UnsupportedFunctionalityError);
    });
  });
});
