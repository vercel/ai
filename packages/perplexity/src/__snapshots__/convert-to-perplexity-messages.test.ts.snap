// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertToPerplexityMessages > assistant messages > should convert an assistant message with text content 1`] = `
[
  {
    "content": "Assistant reply",
    "role": "assistant",
  },
]
`;

exports[`convertToPerplexityMessages > system messages > should convert a system message with text content 1`] = `
[
  {
    "content": "System initialization",
    "role": "system",
  },
]
`;

exports[`convertToPerplexityMessages > user messages > should convert a user message with image parts 1`] = `
[
  {
    "content": [
      {
        "text": "Hello ",
        "type": "text",
      },
      {
        "image_url": {
          "url": "data:image/png;base64,AAECAw==",
        },
        "type": "image_url",
      },
    ],
    "role": "user",
  },
]
`;

exports[`convertToPerplexityMessages > user messages > should convert a user message with text parts 1`] = `
[
  {
    "content": "Hello World",
    "role": "user",
  },
]
`;
